1. Найдите посредством NumPy SVD для матрицы:

$$\begin{pmatrix}
1 & 2 & 0\\ 
0 & 0 & 5\\ 
3 & -4 & 2\\ 
1 & 6 & 5\\ 
0 & 1 & 0
\end{pmatrix}.$$

In [1]:
import numpy as np

In [10]:
np.set_printoptions(precision=2, suppress=True)

In [11]:
A = np.array([[1, 2, 0], [0, 0, 5], [3, -4, 2], [1, 6, 5], [0, 1, 0]])

In [12]:
U, s, W = np.linalg.svd(A)
V = W.T

D = np.zeros_like(A, dtype=float)
D[np.diag_indices(min(A.shape))] = s

In [14]:
print(f'Матрица A =\n{A}\n')
print(f'Матрица U =\n{U}\n')
print(f'Матрица D =\n{D}\n')
print(f'Матрица V =\n{V}\n')

print(f'UDV^T =\n{np.dot(U, np.dot(D, V.T))}\n')

print(f'Сингулярные числа: \n{s}')

Матрица A =
[[ 1  2  0]
 [ 0  0  5]
 [ 3 -4  2]
 [ 1  6  5]
 [ 0  1  0]]

Матрица U =
[[ 0.17  0.16 -0.53 -0.8  -0.16]
 [ 0.39 -0.53  0.61 -0.43  0.03]
 [-0.14 -0.82 -0.52  0.14  0.07]
 [ 0.89  0.06 -0.25  0.38 -0.06]
 [ 0.08  0.11 -0.08 -0.11  0.98]]

Матрица D =
[[8.82 0.   0.  ]
 [0.   6.14 0.  ]
 [0.   0.   2.53]
 [0.   0.   0.  ]
 [0.   0.   0.  ]]

Матрица V =
[[ 0.07 -0.37 -0.93]
 [ 0.72  0.67 -0.21]
 [ 0.69 -0.65  0.31]]

UDV^T =
[[ 1.  2.  0.]
 [ 0. -0.  5.]
 [ 3. -4.  2.]
 [ 1.  6.  5.]
 [-0.  1.  0.]]

Сингулярные числа: 
[8.82 6.14 2.53]


2. Для матрицы из предыдущего задания найдите:

    а) евклидову норму;

Евклидова норма равна максимальному сингулярному числу, т.е:
$$||A||_{E} = \mu_1 = 8.82$$
__Ответ: 8.82__

In [15]:
max(s)

8.824868854820442

    б) норму Фробениуса.

$$||A||_{F} = \sqrt{\sum_{k=1}^r \mu_k^2} = 11.04$$
__Ответ: 11.04__

In [17]:
sum([i*i for i in s]) ** 0.5

11.045361017187261